### PR TITLE
FRR integrated configuration and hitless reloads

### PR DESCRIPTION
**Pull request by bmhughes**
Convert FRR to use an integrated configuration file and use `frr-reload` where possible for hitless configuration changes.

1. Update frr7 port to 7.4.
2. Add the frr7-pythontools port to provide frr-reload.py
3. Refactor the FRR configuration generator to create an integrated configuration file.
3.1. Generated file is in the same format as generated by a vtysh `write mem`.
4. Refactor the rc file/generator to use reload unless a service restart is required. (change of enabled daemons).
5. Accumulate interface descriptions into the FRR configuration for cross-reference to pfSense configuration under `vtysh`
6. Include the `watchfrr` frr watchdog service.
7. Add support for BGP RPKI.

Tested with 2.4.5_p1.

Redmine: https://redmine.pfsense.org/issues/10789

### CODE REVIEW

**Comment by bmhughes** [comment]
Converting to a draft so I can add BGP RPKI support as well.

**Comment by buckweet1980** [comment]
Excited to see this effort to have FRR not do a hard-reload on a config change..

What will it take for this to get pushed out to the pfsense package repository? I assume that's in the works already and it just has to go through a review process?

**Comment by jim-p** [comment]
It needs to be reviewed/revised until accepted, and then it will be available (likely only in 2.5.0 at least initially for testing)

**Comment by bmhughes** [comment]
> 
> 
> Excited to see this effort to have FRR not do a hard-reload on a config change..
> 
> What will it take for this to get pushed out to the pfsense package repository? I assume that's in the works already and it just has to go through a review process?

In the meantime if you're comfortable self-supporting with this and familiar (or prepared to get familiar) with the FreeBSD port package build process you can build the relevant packages yourself easily enough.

**Comment by jim-p** [comment]
Still pondering this, but I did have to make a change to FRR today which caused a conflict that needs resolved. Please rebase and integrate the change from https://github.com/pfsense/FreeBSD-ports/commit/cae5ee237cfabc90ea5ef4dfd480acfc9055e26f
Thanks!

**Comment by bmhughes** [comment]
Done 👍 

Whilst I'm sure there's some room for improvement (I'd like a better method to cache the last used start command for a start), I do think that this is a must-have addition as having the routing protocols reconverge every time you make a change doesn't scale overly well.

**Review by jim-p — CHANGES_REQUESTED** [review]
Seems OK enough to test, but check the leading whitespace in the file I noted.

File: net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_global_raw.xml [hunk 1]
```diff
@@ -2,8 +2,8 @@
 <!DOCTYPE packagegui SYSTEM "../schema/packages.dtd">
 <?xml-stylesheet type="text/xsl" href="../xsl/package.xsl"?>
 <packagegui>
-	<copyright>
-	<![CDATA[
+    <copyright>
+    <![CDATA[
 /*
```
[the remark this answers is not in the document]
Comment (bmhughes) [reply net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_global_raw.xml:7]: Hmm not sure how I've managed that, let me fix it up.

[the remark this answers is not in the document]
Comment (bmhughes) [reply net/pfSense-pkg-frr/files/usr/local/pkg/frr/frr_global_raw.xml]: Converted all the space instances I could find, not sure they were all mine but it's done at least.

**Review by jim-p — APPROVED** [review]

**Comment by rbgarga** [comment]
@bmhughes in order to keep it simple to keep doing future merges I'd prefer to cherry-pick original commits that change `net/frr7` and `net/frr7-pythontools` from FreeBSD.  Could you please remove commits that touch these ports from this pull request and keep only changes to net/pfSense-pkg-frr?

**Comment by rbgarga** [comment]
> @bmhughes in order to keep it simple to keep doing future merges I'd prefer to cherry-pick original commits that change `net/frr7` and `net/frr7-pythontools` from FreeBSD. Could you please remove commits that touch these ports from this pull request and keep only changes to net/pfSense-pkg-frr?

@bmhughes please do not take any action.  I'm manually merging all the changes

**Comment by rbgarga** [comment]
I've manually merged these changes.  Thanks!

**Comment by bmhughes** [comment]
I'm mobile so apologies if you've already covered it but the only thing to be aware of with the frr7 port is the patch I've added from frr-7.4/stable in https://github.com/pfsense/FreeBSD-ports/pull/914/commits/c4614d5e7ab7130f422e92f4cc9161e12def9c93 as that hasn't made it into a release yet to my knowledge and it breaks the reload rather well due to a misformatting of bfd peers from vtysh.

**Comment by rbgarga** [comment]
> I'm mobile so apologies if you've already covered it but the only thing to be aware of with the frr7 port is the patch I've added from frr-7.4/stable in [c4614d5](https://github.com/pfsense/FreeBSD-ports/commit/c4614d5e7ab7130f422e92f4cc9161e12def9c93) as it's been merged but hasn't made it into a release yet to my knowledge and it breaks the reload rather well due to a misformatting of bfd peers from vtysh without it applied.

Yes, I kept these changes.  I just modified the way it replaces ETCDIR adding an entry at make.conf instead of changing net/frr7/Makefile.

I've submitted the changes to net/frr7 port maintainer.  Thank you!

**Pull request closed without merging**